cpp/src/parquet/statistics_test.cc

```diff
@@ -874,160 +875,129 @@ TEST_F(TestStatisticsSortOrderFLBA, UnknownSortOrder) {
-  ASSERT_EQ(min, -4.0f);
-  ASSERT_EQ(max, 3.0f);
-
-  // Test values followed by all NaNs followed by values
-  nan_stats->UpdateSpaced(&values[0], valid_bits.data(), 0, NUM_VALUES, 0);
-  min = nan_stats->min();
-  max = nan_stats->max();
-  ASSERT_EQ(min, -4.0f);
-  ASSERT_EQ(max, 3.0f);
+  auto some_nan_stats = MakeStatistics<ParquetType>(&descr);
+  // Ingesting a mix of NaNs and non-NaNs should not yield valid min max.
+  AssertUnsetMinMax(some_nan_stats, some_nans);
+  // Ingesting only nans should not yield valid min max
+  AssertUnsetMinMax(some_nan_stats, all_nans);
+  // Ingesting NaNs with a null bitmap should not change the result.
+  AssertUnsetMinMax(some_nan_stats, some_nans, &valid_bitmap);
+  // NaNs should not pollute min max when excluded via null bitmap.
+  AssertMinMaxAre(some_nan_stats, some_nans, &valid_bitmap_no_nans, min, max);
 }
 
-// NaN double values may lead to incorrect filtering under certain circumstances
-TEST(TestStatisticsSortOrderDoubleNaN, NaNValues) {
-  constexpr int NUM_VALUES = 10;
-  NodePtr node = PrimitiveNode::Make("nan_double", Repetition::OPTIONAL, Type::DOUBLE);
+TEST(TestStatistic, NaNFloatValues) { CheckNaNs<FloatType>(); }
+
+TEST(TestStatistic, NaNDoubleValues) { CheckNaNs<DoubleType>(); }
+
+// ARROW-7376
+TEST(TestStatisticsSortOrderFloatNaN, NaNAndNullsInfiniteLoop) {
+  constexpr int kNumValues = 8;
+  NodePtr node = PrimitiveNode::Make("nan_float", Repetition::OPTIONAL, Type::FLOAT);
   ColumnDescriptor descr(node, 1, 1);
 
-  auto nan_stats = MakeStatistics<DoubleType>(&descr);
-  double values[NUM_VALUES] = {std::nan(""), std::nan(""), -3.0, -2.0, -1.0,
-                               0.0,          1.0,          2.0,  3.0,  4.0};
-  nan_stats->Update(values, NUM_VALUES, 0);
-  ASSERT_EQ(nan_stats->min(), -3.0);
-  ASSERT_EQ(nan_stats->max(), 4.0);
+  constexpr float nan = std::numeric_limits<float>::quiet_NaN();
+  std::array<float, kNumValues> nans_but_last{nan, nan, nan, nan, nan, nan, nan, 0.0f};
+
+  uint8_t all_but_last_valid = 0x7F;  // 0b01111111
+  auto stats = MakeStatistics<FloatType>(&descr);
+  AssertUnsetMinMax(stats, nans_but_last, &all_but_last_valid);
 }
 
+// ARROW-5562: Ensure that -0.0f and 0.0f values are properly handled like in
+// parquet-mr
 template <typename ParquetType>
 void CheckNegativeZeroStats() {
   using T = typename ParquetType::c_type;
 
-  constexpr int NUM_VALUES = 2;
   NodePtr node = PrimitiveNode::Make("f", Repetition::OPTIONAL, ParquetType::type_num);
   ColumnDescriptor descr(node, 1, 1);
-  T zero = 0;
+  T zero{};
 
-  // Test that the min and max are always unsigned
   {
-    T values[NUM_VALUES] = {-zero, zero};
+    std::array<T, 2> values{-zero, zero};
     auto stats = MakeStatistics<ParquetType>(&descr);
-    stats->Update(values, NUM_VALUES, 0);
-    // The min and max are unsigned
-    ASSERT_FALSE(std::signbit(stats->min()));
-    ASSERT_FALSE(std::signbit(stats->max()));
+    AssertMinMaxAre(stats, values, -zero, zero);
```

Comment:
Equality comparison is not enough to distinguish between negative and positive zero, unfortunately... `AssertMinMaxAre` should probably also compare the `std::signbit`s.

Reply:
+1